vowpalwabbit/core/src/reductions/interaction_ground.cc

```diff
@@ -3,110 +3,279 @@
-  double total_uniform_cost = 0.;
+  ik_stack_builder(VW::LEARNER::base_learner* ftrl_coin) : _ik_base(ftrl_coin)
+  {
+    std::set<std::string> keep = {"scorer", "count_label"};
+
+    for (int i = _reduction_stack.size() - 1; i >= 0; i--) {
+      if (keep.count(std::get<0>(_reduction_stack.at(i))) == 0) {
+        _reduction_stack.erase(_reduction_stack.begin() + i);
+      }
+    }
+  }
+
+  VW::LEARNER::base_learner* setup_base_learner() override {
+    if (_reduction_stack.size() == 0) {
+      return _ik_base;
+    }
+
+    return VW::default_reduction_stack_setup::setup_base_learner();
+  }
+
+private:
+  VW::LEARNER::base_learner* _ik_base;
 };
 
-void negate_cost(VW::multi_ex& ec_seq)
-{
-  for (auto* example_ptr : ec_seq)
-  {
-    for (auto& label : example_ptr->l.cb.costs)
-    {
-      if (label.has_observed_cost()) { label.cost = -label.cost; }
+std::vector<std::vector<VW::namespace_index>> get_ik_interactions(std::vector<std::vector<VW::namespace_index>> interactions, VW::example* observation_ex) {
+  std::vector<std::vector<VW::namespace_index>> new_interactions;
+  for (auto& interaction : interactions) {
+    for (auto& obs_ns : observation_ex->indices) {
+      if (obs_ns == VW::details::DEFAULT_NAMESPACE) {
+        obs_ns = VW::details::IGL_FEEDBACK_NAMESPACE;
+      }
+
+      new_interactions.push_back(interaction);
+      new_interactions.back().push_back(obs_ns);
     }
   }
+
+  return new_interactions;
 }
 
-void learn(interaction_ground& ig, multi_learner& base, VW::multi_ex& ec_seq)
+void add_obs_features_to_ik_ex(VW::example* ik_ex, VW::example* obs_ex) {
+  for (auto& obs_ns : obs_ex->indices) {
+    ik_ex->indices.push_back(obs_ns);
+
+    for (size_t i = 0; i < obs_ex->feature_space[obs_ns].indices.size(); i++) {
+      auto feature_hash = obs_ex->feature_space[obs_ns].indices[i];
+      auto feature_val = obs_ex->feature_space[obs_ns].values[i];
+
+      if (obs_ns == VW::details::DEFAULT_NAMESPACE) {
+        obs_ns = VW::details::IGL_FEEDBACK_NAMESPACE;
+      }
+
+      ik_ex->feature_space[obs_ns].indices.push_back(feature_hash);
+      ik_ex->feature_space[obs_ns].values.push_back(feature_val);
+
+      // ik_ex->num_features++;
+    }
+  }
+}
+
+class interaction_ground
 {
-  // find reward of sequence
-  VW::cb_class label = VW::get_observed_cost_or_default_cb_adf(ec_seq);
-  ig.total_uniform_cost += label.cost / label.probability / ec_seq.size();  //=p(uniform) * IPS estimate
-  ig.total_uniform_reward += -label.cost / label.probability / ec_seq.size();
-
-  // find prediction & update for cost
-  base.predict(ec_seq);
-  ig.total_importance_weighted_cost += get_cost_estimate(label, ec_seq[0]->pred.a_s[0].action);
-  base.learn(ec_seq);
-
-  // find prediction & update for reward
-  label.cost = -label.cost;
-  base.predict(ec_seq, 1);
-  ig.total_importance_weighted_reward += get_cost_estimate(label, ec_seq[0]->pred.a_s[0].action);
+public:
+  VW::LEARNER::single_learner* ik_learner = nullptr;
```

Comment:
```suggestion
  std::shared_ptr<VW::LEARNER::learner> ik_learner = nullptr;
```

This should be enough to fix the memory issues after https://github.com/VowpalWabbit/vowpal_wabbit/pull/4461 gets merged

The other changes that you will probably need to make are
- `base_learner`, `single_learner`, and `multi_learner` all become just `learner`
- `as_multiline` can be deleted or changed to `requires_multiline`, which is just a run-time type check. Same thing for `as_singleline`.
- `make_reduction_learner` and `setup_base_learner` return shared_ptr, so use `auto` instead of `auto*`
```
auto l = make_reduction_learner(...);
return l;
```